core/drivers/crypto/hisilicon/sec_cipher.c

```diff
@@ -0,0 +1,774 @@
+	struct sec_cipher_ctx *c_ctx = msg;
+
+	sqe->bd_param = BD_TYPE3 | SHIFT_U32(SCENE_NOTHING,
+					     SEC_SCENE_OFFSET_V3) |
+			SHIFT_U32(DATA_DST_ADDR_ENABLE, SEC_DE_OFFSET_V3);
+	sqe->c_len_ivin = c_ctx->len;
+	sqe->c_mode_alg = c_ctx->mode |
+			  SHIFT_U32(c_ctx->alg, SEC_CALG_OFFSET_V3);
+	sqe->c_icv_key = SHIFT_U32(c_ctx->c_key_len, SEC_CKEY_OFFSET_V3);
+
+	if (c_ctx->encrypt)
+		sqe->c_icv_key |= CIPHER_ENCRYPT;
+	else
+		sqe->c_icv_key |= CIPHER_DECRYPT;
+
+	sqe->data_dst_addr = c_ctx->out_dma;
+	sqe->data_src_addr = c_ctx->in_dma;
+	sqe->c_key_addr = c_ctx->key_dma;
+	sqe->no_scene.c_ivin_addr = c_ctx->iv_dma;
+
+	return HISI_QM_DRVCRYPT_NO_ERR;
+}
+
+static enum hisi_drv_status sec_cipher_bd3_parse(void *bd, void *msg __unused)
+{
+	struct hisi_sec_bd3_sqe *sqe = bd;
+	uint16_t done = 0;
+
+	done = SEC_GET_FIELD(sqe->done_flag, SEC_DONE_MASK, 0);
+	if (done != SEC_HW_TASK_DONE || sqe->error_type) {
+		EMSG("SEC BD3 fail! done=%#"PRIx16", etype=%#"PRIx8,
+		     done, sqe->error_type);
+		return HISI_QM_DRVCRYPT_IN_EPARA;
+	}
+
+	return HISI_QM_DRVCRYPT_NO_ERR;
+}
+
+static TEE_Result cipher_algo_check(uint32_t algo)
+{
+	switch (algo) {
+	case TEE_ALG_AES_ECB_NOPAD:
+	case TEE_ALG_AES_CBC_NOPAD:
+	case TEE_ALG_AES_CTR:
+	case TEE_ALG_AES_XTS:
+	case TEE_ALG_DES_ECB_NOPAD:
+	case TEE_ALG_DES3_ECB_NOPAD:
+	case TEE_ALG_DES_CBC_NOPAD:
+	case TEE_ALG_DES3_CBC_NOPAD:
+	case TEE_ALG_SM4_CBC_NOPAD:
+	case TEE_ALG_SM4_ECB_NOPAD:
+	case TEE_ALG_SM4_XTS:
+	case TEE_ALG_SM4_CTR:
+		break;
+	default:
+		return TEE_ERROR_NOT_IMPLEMENTED;
+	}
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result crypto_set_alg(struct sec_cipher_ctx *c_ctx, uint32_t alg)
+{
+	TEE_Result ret = TEE_SUCCESS;
+
+	switch (alg) {
+	case TEE_MAIN_ALGO_DES:
+		c_ctx->alg = C_ALG_DES;
+		break;
+	case TEE_MAIN_ALGO_DES3:
+		c_ctx->alg = C_ALG_3DES;
+		break;
+	case TEE_MAIN_ALGO_AES:
+		c_ctx->alg = C_ALG_AES;
+		break;
+	case TEE_MAIN_ALGO_SM4:
+		c_ctx->alg = C_ALG_SM4;
+		break;
+	default:
+		EMSG("Invalid cipher type %#"PRIx32, alg);
+		ret = TEE_ERROR_NOT_IMPLEMENTED;
+		break;
+	}
+
+	return ret;
+}
+
+static TEE_Result crypto_set_mode(struct sec_cipher_ctx *c_ctx, uint32_t mode)
+{
+	TEE_Result ret = TEE_SUCCESS;
+
+	switch (mode) {
+	case TEE_CHAIN_MODE_ECB_NOPAD:
+		c_ctx->mode = C_MODE_ECB;
+		break;
+	case TEE_CHAIN_MODE_CBC_NOPAD:
+		c_ctx->mode = C_MODE_CBC;
+		break;
+	case TEE_CHAIN_MODE_XTS:
+		c_ctx->mode = C_MODE_XTS;
+		break;
+	case TEE_CHAIN_MODE_CTR:
+		c_ctx->mode = C_MODE_CTR;
+		break;
+	default:
+		EMSG("Invalid cipher mode type %#"PRIx32, mode);
+		ret = TEE_ERROR_NOT_IMPLEMENTED;
+		break;
+	}
+
+	return ret;
+}
+
+static TEE_Result sec_cipher_ctx_allocate(void **ctx, uint32_t algo)
+{
+	struct sec_cipher_ctx *c_ctx = NULL;
+	TEE_Result ret = TEE_SUCCESS;
+
+	if (!ctx) {
+		EMSG("ctx is NULL");
+		return TEE_ERROR_STORAGE_NO_SPACE;
+	}
+
+	ret = cipher_algo_check(algo);
+	if (ret)
+		return ret;
+
+	c_ctx = malloc(sizeof(struct sec_cipher_ctx));
```

Comment:
You better use `calloc()` since you don't initialize everything in this function.